### PR TITLE
Skip assemblies that target wrong platform or are non-CLR assemblies

### PR DESCRIPTION
**Pull request by Carlm-MS**

### CODE REVIEW

File: src/Orleans/AssemblyLoader/AssemblyLoader.cs [hunk 1]
```diff
@@ -319,9 +319,16 @@ private bool ReflectionOnlyLoadAssembly(string pathName, out Assembly assembly,
...
                 if (SimulateReflectionOnlyLoadFailure)
                     throw NewTestUnexpectedException();
-                
+
                 assembly = Assembly.ReflectionOnlyLoadFrom(pathName);
             }
+            catch (BadImageFormatException)
+            {
+                // skip assemblies that are don't target the correct platform or are non-clr assemblies.
+                complaints = null;
```
Comment (ReubenBond) [inline src/Orleans/AssemblyLoader/AssemblyLoader.cs]: Would a complaint be useful in this case?

**Comment by ReubenBond** [comment]
One comment, otherwise LGTM

File: src/Orleans/AssemblyLoader/AssemblyLoader.cs (hunk 1 above)
Comment (Carlm-MS) [inline src/Orleans/AssemblyLoader/AssemblyLoader.cs]: I'll update the review. 

It seems that BadImageFormatException in the context of ReflectionOnlyLoad only captures scenarios where a native assemblies is inspected in a managed context.  It does not capture scenarios where a managed assembly is targeting x86 and loaded in a 64-bit process - the reflectionOnlyLoad succeeds but the assembly load fails with a bad image exception.

I'll peek at the PE header to see if the assembly is targeting the correct platform.

Comment (ReubenBond) [inline src/Orleans/AssemblyLoader/AssemblyLoader.cs]: Interesting. You could possibly work out the compatibility by checking both the loaded assembly's `AssemblyName`'s `ProcessorArchitecture` against `Environment.Is64BitProcess`

image

**Comment by attilah** [comment]
Before ae merge it in we need to discuss it IMHO, since this is one of the problems I addressed in my AssemblyLoader work.

**Comment by ReubenBond** [comment]
@attilah should we discuss on this PR then?

**Comment by Carlm-MS** [comment]
The updated commit checks for compatibility of machine type to the currently running executable image before trying to reflection-only load the assembly.

**Comment by attilah** [comment]
@Carlm-MS thanks for updating the PR it looks much consumable without the code that we're getting into the PE and COFF handling stuff.

LGTM.

**Pull request merged**